cpp/arcticdb/processing/clause.hpp

```diff
@@ -39,733 +41,717 @@ namespace arcticdb {
-        return res;
-    }
-
-    [[nodiscard]] const ClauseInfo& clause_info() const {
-        return clause_info_;
-    }
-
-    void set_processing_config(const ProcessingConfig& processing_config) {
-        processing_config_ = processing_config;
-    }
-
-    void set_component_manager(std::shared_ptr<ComponentManager> component_manager) {
-        component_manager_ = component_manager;
-    }
+  [[nodiscard]] std::string to_string() const;
+};
 
-    [[nodiscard]] std::string to_string() const {
-        return fmt::format("GROUPBY Column[\"{}\"]", grouping_column_);
-    }
+template <typename GrouperType, typename BucketizerType> struct PartitionClause {
+  ClauseInfo clause_info_;
+  std::shared_ptr<ComponentManager> component_manager_;
+  ProcessingConfig processing_config_;
+  std::string grouping_column_;
+
+  explicit PartitionClause(const std::string& grouping_column)
+      : processing_config_(), grouping_column_(grouping_column) {
+    clause_info_.input_columns_ = {grouping_column_};
+    clause_info_.requires_repartition_ = true;
+    clause_info_.modifies_output_descriptor_ = true;
+  }
+  PartitionClause() = delete;
+
+  ARCTICDB_MOVE_COPY_DEFAULT(PartitionClause)
+
+  [[nodiscard]] std::vector<std::vector<size_t>>
+  structure_for_processing(std::vector<RangesAndKey>& ranges_and_keys,
+                           size_t start_from) {
+    return structure_by_row_slice(ranges_and_keys, start_from);
+  }
+
+  [[nodiscard]] std::vector<EntityId>
+  process(std::vector<EntityId>&& entity_ids) const {
+    if (entity_ids.empty()) {
+      return {};
+    }
+    auto proc = gather_entities(component_manager_, std::move(entity_ids));
+    std::vector<ProcessingUnit> partitioned_procs =
+        partition_processing_segment<GrouperType, BucketizerType>(
+            proc, ColumnName(grouping_column_), processing_config_.dynamic_schema_);
+    std::vector<EntityId> output;
+    for (auto&& partitioned_proc : partitioned_procs) {
+      std::vector<EntityId> proc_entity_ids =
+          push_entities(component_manager_, std::move(partitioned_proc));
+      output.insert(output.end(), proc_entity_ids.begin(), proc_entity_ids.end());
+    }
+    return output;
+  }
+
+  [[nodiscard]] std::optional<std::vector<std::vector<EntityId>>>
+  repartition(std::vector<std::vector<EntityId>>&& entity_ids_vec) const {
+    schema::check<ErrorCode::E_COLUMN_DOESNT_EXIST>(
+        std::any_of(entity_ids_vec.cbegin(), entity_ids_vec.cend(),
+                    [](const std::vector<EntityId>& entity_ids) {
+                      return !entity_ids.empty();
+                    }),
+        "Grouping column {} does not exist or is empty", grouping_column_);
+    // Some could be empty, so actual number may be lower
+    auto max_num_buckets = ConfigsMap::instance()->get_int(
+        "Partition.NumBuckets", async::TaskScheduler::instance()->cpu_thread_count());
```

Comment:
I personally find this harder to read than when the closing parentheses are on a newline

Reply:
Me too

Reply:
I can't find the option for this one. Please suggest a change to the ruleset

Reply:
I think it's a pity to enforce them since there is a tradeoff between clarity and wasted vertical space. Particularly since compilers actually warn/fail on misleading indentiation these days. On the other hand, I am in favour of using unbraced statements very sparingly, i.e. only when there is a simple if statement and a single line for both if and else clauses. Any multi-line blocks, multiple else-if statements or any other kind of complexity should mean that everything is entirely braced